flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp

```diff
@@ -99,87 +75,360 @@ class DoConcurrentConversion : public mlir::OpConversionPattern<fir::DoLoopOp> {
+          // We want to collect ops corresponding to live-ins only once.
+          if (definingOp && !seenOps.insert(definingOp).second)
+            return;
+
+          liveIns.push_back(operand->get());
+
+          if (isIndVarUltimateOperand(operand->getOwner()))
+            std::swap(*liveIns.begin(), *liveIns.rbegin());
+        });
+    return liveIns;
+  }
+
+  mlir::omp::MapInfoOp
+  genMapInfoOpForLiveIn(mlir::ConversionPatternRewriter &rewriter,
+                        mlir::Value liveIn) const {
+    hlfir::DeclareOp declareOp =
+        mlir::dyn_cast_if_present<hlfir::DeclareOp>(liveIn.getDefiningOp());
+
+    if (declareOp == nullptr)
+      TODO(liveIn.getLoc(),
+           "Values not defined by declare op's are not supported yet.");
+
+    mlir::Type liveInType = liveIn.getType();
+    mlir::Type eleType = liveInType;
+    if (auto refType = liveInType.dyn_cast<fir::ReferenceType>())
+      eleType = refType.getElementType();
+
+    llvm::omp::OpenMPOffloadMappingFlags mapFlag =
+        llvm::omp::OpenMPOffloadMappingFlags::OMP_MAP_IMPLICIT;
+    mlir::omp::VariableCaptureKind captureKind =
+        mlir::omp::VariableCaptureKind::ByRef;
+
+    if (fir::isa_trivial(eleType) || fir::isa_char(eleType)) {
+      captureKind = mlir::omp::VariableCaptureKind::ByCopy;
+    } else if (!fir::isa_builtin_cptr_type(eleType)) {
+      mapFlag |= llvm::omp::OpenMPOffloadMappingFlags::OMP_MAP_TO;
+      mapFlag |= llvm::omp::OpenMPOffloadMappingFlags::OMP_MAP_FROM;
     }
 
-    // For each collected `fir.store`, find the target memref's alloca's and
-    // declare ops.
-    llvm::SmallSetVector<mlir::Operation *, 4> declareAndAllocasToClone;
-    for (auto storeOp : inductionVarTargetStores) {
-      mlir::Operation *storeTarget = storeOp.getMemref().getDefiningOp();
+    llvm::SmallVector<mlir::Value> boundsOps = [&]() {
+      llvm::SmallVector<mlir::Value> result;
 
-      for (auto operand : storeTarget->getOperands()) {
-        declareAndAllocasToClone.insert(operand.getDefiningOp());
+      if (declareOp.getShape() == nullptr) {
+        return result;
       }
-      declareAndAllocasToClone.insert(storeTarget);
+
+      fir::ShapeOp shapeOp = mlir::dyn_cast_if_present<fir::ShapeOp>(
+          declareOp.getShape().getDefiningOp());
+
+      if (shapeOp == nullptr)
+        TODO(liveIn.getLoc(),
+             "Shapes not defined by shape op's are not supported yet.");
+
+      auto extents = shapeOp.getExtents();
+
+      auto genBoundsOp = [&](mlir::Value extent) {
+        mlir::Type extentType = extent.getType();
+        mlir::arith::ConstantOp lb = rewriter.create<mlir::arith::ConstantOp>(
+            liveIn.getLoc(), extentType,
+            rewriter.getIntegerAttr(extentType, 0));
+        // TODO I think this caluclation might not be correct. But this is how
+        // it is done in PFT->OpenMP lowering. So keeping it like this until we
+        // double check.
+        mlir::Value ub =
```

Comment:
You can likely verify the calculation in the DirectiveCommons.h code where the bounds are generated, and if still in doubt fire a message towards one of the OpenACC guys who created it, there's been a bug or two before, but they're generally quite good at clarifying things :-)

But I suppose if the lower bound is 0, and the extent is the max size of a dimension, you'd get the upper bound, but the subtraction seems like extra legwork at a minimum here I think, as you could just take the extent I would imagine.